internal/postrs/bridge.go

```diff
@@ -0,0 +1,266 @@
+package postrs
+
+// #cgo LDFLAGS: -lpost
+// #include <stdlib.h>
+// #include "prover.h"
+import "C"
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"reflect"
+	"sync"
+	"unsafe"
+
+	"github.com/spacemeshos/post/config"
+	"github.com/spacemeshos/post/shared"
+)
+
+// gpuMtx is a mutual exclusion lock for calls to gpu functions. It is required
+// to prevent concurrent calls to the GPU library that are expected to cause a
+// crash.
+var gpuMtx sync.Mutex
+
+type DeviceClass int
+
+const (
+	ClassCPU     = DeviceClass((C.DeviceClass)(C.CPU))
+	ClassGPU     = DeviceClass((C.DeviceClass)(C.GPU))
+	ClassUnknown = DeviceClass((C.DeviceClass)(C.UNKNOWN))
+)
+
+type ComputeProvider struct {
+	ID         uint
+	Model      string
+	DeviceType DeviceClass
+}
+
+func (c DeviceClass) String() string {
+	switch c {
+	case ClassCPU:
+		return "CPU"
+	case ClassGPU:
+		return "GPU"
+	default:
+		return "Unknown"
+	}
+}
+
+var (
+	ErrInvalidProviderID = errors.New("invalid provider ID")
+
+	ErrInvalidLabelsRange = errors.New("invalid labels range")
+	ErrOclError           = errors.New("OpenCL error")
+	ErrInvalidArgument    = errors.New("invalid argument")
+	ErrFetchProviders     = errors.New("failed to fetch providers")
+)
+
+const (
+	// LabelLength is the length of the label in bytes.
+	LabelLength = 16
+)
+
+func InitResultToError(retVal uint32) error {
+	switch retVal {
+	case C.InitializeOk:
+		return nil
+	case C.InitializeInvalidLabelsRange:
+		return ErrInvalidLabelsRange
+	case C.InitializeOclError:
+		return ErrOclError
+	case C.InitializeInvalidArgument:
+		return ErrInvalidArgument
+	case C.InitializeFailedToGetProviders:
+		return ErrFetchProviders
+	default:
+		return fmt.Errorf("unknown error")
+	}
+}
+
+func cScryptPositions(opt *option) ([]byte, *uint64, error) {
+	gpuMtx.Lock()
+	defer gpuMtx.Unlock()
```

Comment:
I would like to use a `TryLock()` here to return early with an error ("device busy"), but this breaks a whole lot of tests in `go-spacemesh` that are running in parallel (and probably shouldn't).